src/electionguard/async.hpp

```diff
@@ -0,0 +1,324 @@
+#ifndef __ELECTIONGUARD_CPP_ASYNC_HPP_INCLUDED__
+#define __ELECTIONGUARD_CPP_ASYNC_HPP_INCLUDED__
+
+#include "electionguard/export.h"
+//#include "thread_pool.hpp"
+#include "log.hpp"
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+
+using std::atomic;
+using std::condition_variable;
+using std::future;
+using std::future_status;
+using std::mutex;
+using std::unique_lock;
+using std::unique_ptr;
+using std::vector;
+using std::chrono::seconds;
+
+namespace electionguard
+{
+    /// <summary>
+    /// A simple locking mechanism that allows the compiler to choose
+    /// how to implement the deferral.
+    ///
+    /// use like this:
+    /// AsyncSemaphore task_lock(4); //<-- count of threads
+    /// auto task = std::async(std::launch::async,
+    ///                        [](AsyncSemaphore &mutex) {
+    ///                            mutex.wait();
+    ///                            return doStuff();
+    ///                        },
+    ///                        std::ref(task_lock)));
+    /// </summary>
+    class EG_INTERNAL_API AsyncSemaphore
```

Comment:
this isnt currently used in the project but it will be in the future.  I used it for testing initially and it is a way to do an async lock without requiring the use of the full threadpool